File: Medium/Python/1833. Maximum Ice Cream Bars.py
class Solution(object):
    def maxIceCream(self, costs, coins):
        # Counting Sort Algorithm
        # Find the max in the array
        maximum = costs[0]
        for c in costs:
            maximum = max(c, maximum)
        
        # Find the frequency of each cost
        costFreq = [0] * (maximum + 1)
        for c in costs:
            costFreq[c] += 1

        # Buy the lowest cost ice cream bars to maximize
        # amount of ice cream bars purchased
        count = 0
        for i in range(1, maximum + 1):
            freq = costFreq[i]
            if coins < i:
                break
            buy = min(freq, coins / i)
            coins -= i * buy
            count += buy
        
        return count

File: Medium/JavaScript/1833. Maximum Ice Cream Bars.js
var maxIceCream = function(costs, coins) {
    // Counting Sort Algorithm
    // Find the max in the array
    let maximum = costs[0];
    costs.forEach((c) => {
        maximum = Math.max(c, maximum);
    });
    
    // Find the frequency of each cost
    costFreq = Array(maximum + 1).fill(0);
    costs.forEach((c) => {
        costFreq[c]++;
    })
    
    // Buy the lowest cost ice cream bars to maximize
    // the amount of ice cream bars purchased
    let count = 0;
    for(let i = 1; i <= maximum; i++){
        let freq = costFreq[i];
        if(coins < i){
            break;
        }
        let buy = Math.min(freq, Math.floor(coins / i));
        coins -= i * buy;
        count += buy;
    }

    return count;
};
